File: atm-6/java/BruteForceIE.java
//package net.nadschlaeger;

import java.util.List;
import java.util.ArrayList;

/**
 * Created by snadschlaeger on 28.12.2016.
 */
public class BruteForceIE extends AbstractIE {

	List<Object> rules;
	List<Object> dbFacts;
	List<Object> rtFacts;

	public BruteForceIE(List<Object> rules, List<Object> facts) {
		this.rules = rules;
		dbFacts = facts;
	}

	@Override
	protected void assertFact(Object fact) {
		rtFacts.add(fact);
	}

	@Override
	protected void clearRuntimeFactStorage() {
		rtFacts = new ArrayList<>();
	}

	@Override
	protected List<Object> getAllFacts() {
		return dbFacts;
	}

	@Override
	protected List<Object> getAllFactsInRuntimeStorage() {
		return rtFacts;
	}

	@Override
	protected List<Object> getAllRules() {
		return rules;
	}

	@Override
	protected String getConditionExpression(Object condition) {
		return ((Clause)condition).op.name();
	}

	@Override
	protected Object getConditionParameter(Object condition) {
		return ((Clause)condition).fact;
	}

	@Override
	protected String getConsequenceExpression(Object consequence) {
		return ((Consequence)consequence).op.name();
	}

	@Override
	protected Object getConsequenceParameter(Object consequence) {
		return ((Consequence)consequence).parameter;
	}

	@Override
	protected String getFactName(Object fact) {
		return ((Fact)fact).name;
	}

	@Override
	protected List<Object> getFactProperties(Object fact) {
		return ((Fact)fact).properties;
	}

	@Override
	protected String getFactPropertyName(Object property) {
		return ((Property)property).name;
	}

	@Override
	protected Object getFactPropertyOnPosition(Object fact, int position) {
		return ((Fact)fact).properties.get(position);
	}

	@Override
	protected String getFactPropertyType(Object property) {
		return ((Property)property).type;
	}

	@Override
	protected Object getFactPropertyValue(Object property) {
		return ((Property)property).value;
	}

	@Override
	protected int getNrOfFactProperties(Object fact) {
		return ((Fact)fact).properties.size();
	}

	@Override
	protected List<Object> getRuleConditions(Object rule) {
		return ((Rule)rule).conditions;
	}

	@Override
	protected List<Object> getRuleConsequences(Object rule) {
		return ((Rule)rule).consequences;
	}

	@Override
	protected RuleOperator getRuleOperator(Object rule) {
		return ((Rule)rule).op;
	}

	@Override
	protected boolean isFactPropertyMandatory(Object property) {
		// TODO Auto-generated method stub
		return false;
	}

	@Override
	protected boolean propertyRequiresInputFromUser(Object property) {
		return (getFactPropertyValue(property) == null);
	}

	@Override
	protected void setConditionParameter(Object clause, Object parameter) {
		((Clause) clause).fact = (Fact)parameter;
	}

	@Override
	protected void setFactPropertyValue(Object property, Object value) {
		((Property)property).value = value;
	}

}

File: atm-8/ATM.Logic/BruteForceIE.cs
using System;
using System.Collections.Generic;
using System.Linq;
using ATM.Domain;


namespace ATM.Logic
{
    public class BruteForceIE : AbstractIE
    {

        public List<object> RuntimeFacts { get; set; }

        protected override void AssertFact(object fact)
        {
            throw new NotImplementedException();
        }

        protected override void ClearRuntimeFactStorage()
        {
            RuntimeFacts.Clear();
        }

        protected override List<object> GetAllFacts()
        {
            return CardStore.GetAll();
        }

        protected override List<object> GetAllFactsInRuntimeStorage()
        {
            return RuntimeFacts;
        }

        protected override string GetConditionExpression(object condition)
        {
            return ((Condition)condition).Expression;
        }

        protected override object GetConditionParameter(object condition)
        {
            return ((Condition)condition).Parameter;
        }

        protected override string GetConsequenceExpression(object consequence)
        {
            return ((Domain.Action) consequence).expression;
        }

        protected override object GetConsequenceParameter(object consequence)
        {
            return ((Domain.Action)consequence).Parameter;
        }

        protected override string GetFactName(object fact)
        {
            return ((CreditCard)fact).CardNumber.ToString();
        }

        protected override List<object> GetFactProperties(object fact)
        {
            var lst = new List<object>();
            if(fact.GetType() == typeof(CreditCard))
            {
                lst.Add(((CreditCard)fact).CardNumber);
            }
            else if (fact.GetType() == typeof(Pin))
            {
                lst.Add(((Pin)fact).CardNumber);
                lst.Add(((Pin)fact).PinCode);
            }
            else if (fact.GetType() == typeof(Account))
            {
                lst.Add(((Account)fact).CardNumber);
                lst.Add(((Account)fact).Balance);
            }

            return lst;
        }

        protected override string GetFactPropertyName(object property)
        {
            return nameof(property);
        }

        protected override object GetFactPropertyOnPosition(object fact, int position)
        {
            var props = GetFactProperties(fact);
            return props.ElementAt(position);

        }

        protected override string GetFactPropertyType(object property)
        {
            return property.GetType().FullName;
        }

        protected override object GetFactPropertyValue(object property)
        {
            throw new NotImplementedException();
        }

        protected override int GetNrOfFactProperties(object fact)
        {
            return GetFactProperties(fact).Count;
        }

        protected override List<object> GetAllRules()
        {
            return RuleStore.Rules;
        }

        protected override RuleOperator GetRuleOperator(object rule)
        {
            return ((Rule)rule).RuleOperator;
        }


        protected override List<object> GetRuleConditions(object rule)
        {
            return ((Rule)rule).Conditions;
        }

        protected override List<object> GetRuleConsequences(object rule)
        {
            return ((Rule) rule).Actions;
        }

        
        protected override bool IsFactPropertyMandatory(object property)
        {
            throw new NotImplementedException();
        }

        protected override bool PropertyRequiresInputFromUser(object property)
        {
            throw new NotImplementedException();
        }

        protected override void SetConditionParameter(object condition, object parameter)
        {
            ((Condition)condition).Parameter = parameter;
        }

        protected override void SetFactPropertyValue(object property, object value)
        {
            
        }
    }
}
